libarchive/archive_read_disk_posix.c

```diff
@@ -1290,7 +1290,7 @@ archive_read_disk_descend(struct archive *_a)
 	    ARCHIVE_STATE_HEADER | ARCHIVE_STATE_DATA,
 	    "archive_read_disk_descend");
 
-	if (t->visit_type != TREE_REGULAR || !t->descend)
+	if (!archive_read_disk_can_descend(_a))
```

Comment:
I think there is a problem in the archive_read_disk_can_descend() function. It calls archive_check_magic() which can return ARCHIVE_FATAL which becomes a return value of  archive_read_disk_can_descend(). As it converted to bool that will cause the rest of the function to be skipped in case of bad magic. If the check above is removed as redundant in the future - this can lead to behavior change. Ideally, archive_read_disk_can_descend() should return ARCHIVE_OK or other statuses instead of bool and checked appropriately at callers. There are several call sites where there is no archive_check_magic() before invocation of archive_read_disk_can_descend().

Reply:
`archive_check_magic` is a sanity check. If it triggers, the caller is doing something completely wrong.

Reply:
The problem is bad interface - the archive_read_disk_can_descend() return BOOL and it is true when the sanity check fails.

Reply:
Again, irrelevant. It's checked already just above this line.